### PR TITLE
Alan Ren PR

### PR DESCRIPTION
**Pull request by alanren2**

### CODE REVIEW

File: src/components/Counter.component.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+class Counter extends Component {
+  state = { count: this.props.initialState }
```
Comment (CodeBrew28) [inline src/components/Counter.component.js:5]: Wrap this in a constructor, so like
```
constructor(props) {
     super(props)
     this.state = { count: this.props.initialState }
}
```

File: src/components/Counter.component.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  handleUp = () => {
+    this.setState({ count: this.state.count + 1 })
+  }
+  handleDown = () => {
```
Comment (CodeBrew28) [inline src/components/Counter.component.js:9]: Nice! You used arrow functions :)

File: src/components/Instructions.component.js [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
+import PropTypes from 'prop-types'
 import './../styles/instructions.css'
 
+function moreThan3(str) {
+  return str.length > 3
+}
 class Instructions extends Component {
   render() {
+    const ListItems = this.props.items
+      .filter(moreThan3)
```
Comment (CodeBrew28) [inline src/components/Instructions.component.js:11]: you can do this in one line instead of creating a moreThan3 function 
```
.filter(str => str.length > 3)
```

File: src/components/InitialCountForm.component.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+class InitialCountForm extends Component {
+  state = { submitted: false, valid: true }
```
Comment (CodeBrew28) [inline src/components/InitialCountForm.component.js:5]: same as above, use a constructor for initial state

File: src/components/Counter.component.js (hunk 1 above)
Comment (vrunjeti) [reply src/components/Counter.component.js:5]: actually, this is more of a matter of style, but it's perfectly fine to leave it as is. i personally prefer to not create constructors if they're not needed

File: src/components/InitialCountForm.component.js (hunk 4 above)
Comment (vrunjeti) [reply src/components/InitialCountForm.component.js:5]: same here

File: src/App.js [hunk 5]
```diff
@@ -1,13 +1,40 @@
 import React, { Component } from 'react'
-import { Instructions } from './components'
+import { Instructions, InitialCountForm, Counter } from './components'
 import './styles/app.css'
 
 class App extends Component {
+  state = { submitted: true, valid: true, count: 0 }
+  displayCounter = event => {
+    event.preventDefault()
+    let input = event.target[0].value
```
Comment (vrunjeti) [inline src/App.js:9]: use `const` unless you need to reassign variables

https://github.com/airbnb/javascript#references--prefer-const

File: src/App.js [hunk 6]
```diff
@@ -1,13 +1,40 @@
...
+        />
+        <InitialCountForm handler={this.displayCounter} />
+        {this.state.countSubmitted ? (
+          this.state.invalidInput ? (
+            <h3>Please enter a number</h3>
+          ) : (
+            <Counter initialState={this.state.count} />
+          )
+        ) : (
+          <div />
```
Comment (vrunjeti) [inline src/App.js:36]: avoid using nested ternary operators, it gets hard to follow. 

for the outside one, since your "else" condition is an empty div, there's no need to render it at all. if you only need to render the "then" clause of a conditional, you can use `&&` like so:

```jsx
this.state.countSubmitted && (JSX)
```

File: src/components/InitialCountForm.component.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  displayCounter = event => {
```
Comment (vrunjeti) [inline src/components/InitialCountForm.component.js:6]: good job on moving this logic to the parent component. you should delete it here though

File: src/components/index.js [hunk 8]
```diff
@@ -1,3 +1,7 @@
 import Instructions from './Instructions.component'
+import Counter from './Counter.component'
+import InitialCountForm from './InitialCountForm.component'
 
 export { Instructions }
+export { Counter }
+export { InitialCountForm }
```
Comment (vrunjeti) [inline src/components/index.js:7]: you can single line export these: 

```jsx
export { Instructions, Counter, InitialCountForm }
```

File: src/components/Counter.component.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    return (
+      <div>
+        Count:
+        <data> {count} </data>
+        <button onClick={this.handleUp}> Increment </button>
+        <button onClick={this.handleDown}> Decrement </button>
+      </div>
+    )
+  }
+}
```
Comment (vrunjeti) [inline src/components/Counter.component.js:23]: so right now, if you update the initial count input, it won't update in this counter component. there are a couple ways of fixing that:

1. using [`componentWillReceiveProps`](https://reactjs.org/docs/react-component.html#componentwillreceiveprops)
2. moving the counter state and state updating functions up to the parent component

the former will register changes in incoming props which you can use to update your components state, but the latter will allow your counter component to not have any state at all (as it will live in the `App` component). that's preferred since you want to try to minimize state as much as possible. `Counter` will then become a stateless component, which is easier to reason about since its `render` is only a function of its `props`.

https://flaviocopes.com/react-beginners-guide/#moving-the-state-up-in-the-tree

File: src/components/InitialCountForm.component.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  }
+  render() {
+    return (
+      <form onSubmit={this.props.handler}>
+        <label>
+          Input Initial Count:
+          <input name="Initial Count: " type="number" />
+        </label>
+        <button type="submit">Submit</button>
+        <br />
```
Comment (vrunjeti) [inline src/components/InitialCountForm.component.js:27]: not sure what you put this in for, but its unnecessary. if you used it for spacing, you shouldn't use html markup to do so, you should use css

**Review by vrunjeti — CHANGES_REQUESTED** [review]
you have the right idea going, but you can improve this by restructuring the application state a bit. pulling up state to the top will minimize state and simplify the app. aside from that, there's a few stylistic changes you can make
